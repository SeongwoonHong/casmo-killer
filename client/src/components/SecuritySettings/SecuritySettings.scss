.Security-settings {
    &__form {
        > .user-form-fields {
            > .user-form-button {
                &:last-child {
                    margin: $user-box-padding 0 0 auto;
                }
            }
        }
        > .button-group {
            @include flex-box(space-between);
            padding: $user-box-padding;
            > .user-form-button {
                cursor: pointer;
                width: 100%;
                &:first-child {
                    margin-right: $user-box-padding / 2;
                    &:disabled {
                        background-color: #cdcdcd;
                        cursor: not-allowed;
                        pointer-events: none;
                    }
                }
                &:last-child {
                    margin-left: $user-box-padding / 2;
                    text-decoration: none;
                }
            }
        }
    }
}
