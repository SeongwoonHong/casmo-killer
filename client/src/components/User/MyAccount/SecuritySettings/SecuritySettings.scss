.Security-settings {
    > .user-form {
        > .user-form-fields {
            > .user-form-button {
                &:last-child {
                    margin: $user-box-padding 0 0 auto;
                }
            }
        }
        > .button-group {
            @include flex-box(space-between);
            padding: $user-box-padding;
            > .user-form-button {
                width: 100%;
                &:first-child {
                    margin-right: $user-box-padding / 2;
                }
                &:last-child {
                    margin-left: $user-box-padding / 2;
                }
            }
        }
    }
}
