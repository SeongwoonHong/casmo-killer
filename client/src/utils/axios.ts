import axios from 'axios';

axios.interceptors.request.use((config) => {
  console.log(config);
  return config;
});

function setTokenToHeader(header, csrfToken) {
  axios.defaults.headers.common[header] = csrfToken;
}

export {
  axios,
  setTokenToHeader,
};
