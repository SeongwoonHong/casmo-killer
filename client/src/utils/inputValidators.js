import inputValidator from 'validator';
import axios from 'axios';

export const trim = (input) => {
  return inputValidator.trim(input);
};

export const validateImg = (img) => {
  return inputValidator.isDataURI(img) || inputValidator.isURL(img);
};

export const validateEmail = async (email) => {

  if (inputValidator.isEmpty(email)) {
    return 'Please enter email address.';
  } else if (/\s/g.test(email)) {
    return 'Email address cannot have any space.';
  } else if (!inputValidator.isEmail(email)) {
    return 'The provided email address is not valid.';
  }

  try {
    const { data } = await axios.get(`/api/auth/verify/email/${email}`);
    if (!data) {
      return 'Internal Server Error.';
    }
    if (data.isDuplicate) {
      return 'The email address is already registered.';
    }
  } catch (error) {
    console.error(error);
    return 'Internal Server Error.';
  }

  return '';

};

export const validateDisplayName = async (displayName) => {

  if (inputValidator.isEmpty(displayName)) {
    return 'Please enter display name.';
  } else if (displayName.length < 4) {
    return 'Display name must be more than 4 characters.';
  } else if (displayName.length > 20) {
    return 'Display name must be less than 20 characters.';
  } else if (/\s/g.test(displayName)) {
    return 'Display name cannot have any space.';
  } else if (!/^[a-zA-Z0-9ㄱ-ㅎ가-힣]{4,20}/.test(displayName)) {
    return 'Display name cannot have special characters.';
  }

  try {
    const { data } = await axios.get(`/api/auth/verify/displayName/${displayName}`);
    if (!data) {
      return 'Internal Server Error.';
    }
    if (data.isDuplicate) {
      return 'The display name is already registered.';
    }
  } catch (error) {
    console.error(error);
    return 'Internal Server Error.';
  }

  return '';

};

export const validatePassword = async (password, confirmPassword = null, isReset = false) => {

  // TODO: password may need more robust validation
  if (inputValidator.isEmpty(password)) {
    return 'Please enter password.';
  } else if (password.length < 6) {
    return 'Password must be more than 6 characters.';
  } else if (password.length > 20) {
    return 'Password must be less than 20 characters.';
  } else if (confirmPassword !== null) {
    if (password !== confirmPassword) {
      return 'Password does not match the confirm password.';
    } else if (password === confirmPassword && !isReset) {
      try {
        const { status } = await axios.post('/api/user/verify/password', { password });
        if (status === 204) {
          return 'New password must be different from current password.';
        }
      } catch (error) {
        console.error(error);
      }
    }

  }

  return '';

};
