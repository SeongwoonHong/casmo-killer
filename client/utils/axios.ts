import axios from 'axios';

const baseURL = 'http://localhost:9000';

axios.defaults.baseURL = baseURL + '/api';

export { axios };
