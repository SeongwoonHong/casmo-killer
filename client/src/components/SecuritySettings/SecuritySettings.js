import React, { Component } from 'react';
import axios from 'axios';
import { Link } from 'react-router-dom';

import { validatePassword } from '@sharedUtils/inputValidators';

import FormMessage from '@sharedComponents/FormMessage';

import UserPageContainer from '../UserPageContainer';
import UserInputField from '../UserInputField';

import './SecuritySettings.scss';

const initialState = {
  isLoading: false,
  isEditing: false,
  isVerified: false,
  isSuccess: false,
  currentPassword: '',
  newPassword: '',
  confirmPassword: '',
  message: ''
};

class SecuritySettings extends Component {

  constructor(props) {

    super(props);

    this.state = initialState;

  }

  onChangeHandler = (e) => {
    this.setState({ [e.name]: e.value });
  };

  onSubmitHandler = () => {

    this.setState({
      isLoading: true,
      message: ''
    });

    if (!this.state.isVerified) {
      this.verifyCurrentPassword();
    } else {
      this.submitNewPassword();
    }

  };

  verifyCurrentPassword = async () => {

    const { currentPassword } = this.state;

    if (currentPassword.length === 0) {

      this.setState({
        isSuccess: false,
        message: 'Please enter your password.'
      });

    } else {

      try {

        const { status } = await axios.post('/api/user/verify/password', {
          password: currentPassword
        });

        if (status === 204) {

          this.setState(Object.assign({}, initialState, {
            isVerified: true
          }));

        } else {

          this.setState({
            isSuccess: false,
            message: 'Failed to communicate with the server.'
          });

        }

      } catch (error) {

        console.error(error);
        this.setState({
          isSuccess: false,
          message: error.response.data.message
        });

      }

    }

    this.setState({ isLoading: false });

  };

  submitNewPassword = async () => {

    const { newPassword, confirmPassword } = this.state;

    const validationMsg = await validatePassword(newPassword, confirmPassword);

    if (validationMsg.length > 0) {

      this.setState({
        isSuccess: false,
        message: validationMsg
      });

    } else {

      try {

        const { status } = await axios.put('/api/user/update/password', { newPassword });

        if (status === 204) {
          this.setState({
            isSuccess: true,
            currentPassword: '',
            newPassword: '',
            confirmPassword: '',
            message: 'Your password has been successfully updated.'
          });
        } else {
          this.setState({
            isSuccess: false,
            message: 'Failed to communicate with the server.'
          });
        }

      } catch (error) {

        console.error(error.response.data.error);
        this.setState({
          isSuccess: false,
          message: error.response.data.message
        });

      }

    }

    this.setState({ isLoading: false });

  };

  render() {

    const {
      isLoading,
      isEditing,
      isVerified,
      isSuccess,
      currentPassword,
      newPassword,
      confirmPassword,
      message,
    } = this.state;

    const { strategy } = this.props;

    return (
      <UserPageContainer
        className="Security-settings"
        isLoading={ isLoading }
        formTitle="Security Settings"
        formMsg="Change your password, or delete your account"
        onSubmit={ this.onSubmitHandler }
        button={ (
          <div className="button-group">
            <button
              type="button"
              disabled={ strategy }
              className="user-form-button"
              onClick={ () => {
                if (isVerified) {
                  this.setState(initialState);
                } else if (!isEditing) {
                  this.setState({ isEditing: true });
                }
              }}>
              { isEditing || isVerified ? 'Cancel' : 'Change Password' }
            </button>
            <Link
              to="/user/delete"
              className="user-form-button">
              Delete Account
            </Link>
          </div>
        ) }>

        <FormMessage
          message={ message }
          type={ isSuccess ? 'success' : 'error' } />

        <UserInputField
          isVisible={ this.state.isEditing }
          type="password"
          name="currentPassword"
          title="Current Password"
          onChange={ this.onChangeHandler }
          value={ currentPassword }
          message="Please enter your current password to proceed.">
          <button
            type="submit"
            className="user-form-button">
            Submit
          </button>
        </UserInputField>

        <UserInputField
          isVisible={ this.state.isVerified }
          type="password"
          name="newPassword"
          title="New Password"
          onChange={ this.onChangeHandler }
          value={ newPassword }
          message="Enter new password" />
        <UserInputField
          isVisible={ this.state.isVerified }
          type="password"
          name="confirmPassword"
          title="Confirm Password"
          onChange={ this.onChangeHandler }
          value={ confirmPassword }
          message="Confirm your password">
          <button
            type="submit"
            className="user-form-button">
            Submit
          </button>
        </UserInputField>

      </UserPageContainer>
    );

  }

}

export default SecuritySettings;
