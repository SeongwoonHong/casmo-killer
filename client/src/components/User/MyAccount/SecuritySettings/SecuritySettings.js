import React, { Component } from 'react';
import axios from 'axios';
import PlainBtn from 'sharedComponents/PlainBtn';
import LoadingCircle from 'sharedComponents/LoadingCircle';

import inputValidator from 'sharedUtils/inputValidator';

import './SecuritySettings.scss';

class SecuritySettings extends Component {

  constructor(props) {
    super(props);

    this.state = {
      isEditing: false,
      isLoading: false,
      hasBeenVerified: false,
      currentPassword: '',
      newPassword: '',
      confirmPassword: '',
      message: []
    };
  }

  onChangeHandler = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  onSubmitHandler = (e) => {
    e.preventDefault();
    if (!this.state.hasBeenVerified) {
      this.setState({
        isEditing: true
      });
    }
  };

  verifyCurrentPassword = async () => {

    this.setState({ isLoading: true });

    try {

      const { status } = await axios.post('/api/user/validate/password', {
        password: this.state.currentPassword
      });

      if (status === 204) {
        this.setState({
          isLoading: false,
          isEditing: false,
          hasBeenVerified: true,
          message: []
        });
      }

    } catch (error) {

      this.setState({
        message: [error.response.data.message]
      });

    }

    this.setState({ isLoading: false });

  };

  submitNewPassword = async () => {

    this.setState({ isLoading: true });

    const {
      currentPassword,
      newPassword,
      confirmPassword
    } = this.state;

    const messages = [];

    this.setState({ message: [] });

    if (inputValidator.isEmpty(newPassword)) {
      messages.push('Please enter new password.');
    } else if (newPassword.length < 6) {
      messages.push('Password must be more than 6 characters.');
    } else if (newPassword.length > 20) {
      messages.push('Password must be less than 20 characters.');
    }

    if (inputValidator.isEmpty(confirmPassword)) {
      messages.push('Please confirm new password.');
    } else if (newPassword !== confirmPassword) {
      messages.push('Password does not match the confirm password.');
    }

    if (newPassword === currentPassword) {
      messages.push('New password must be different from current password.');
    }

    if (messages.length > 0) {
      return this.setState({
        isLoading: false,
        message: messages
      });
    }

    try {

      const { status } = await axios.put('/api/user/update/password', { newPassword });

      if (status === 204) {
        this.setState({
          isEditing: false,
          isLoading: false,
          hasBeenVerified: false,
        });
        this.props.onPasswordChange();
      }

    } catch (error) {

      console.error(error.response.data.error);
      this.setState({
        isLoading: false,
        message: [error.response.data.message]
      });

    }

  };

  render() {
    return (
      <div className="card">
        <form
          noValidate
          onSubmit={ this.onSubmitHandler }
          className="for-security">
          <div className="settings-header">
            <h4>Security Settings</h4>
            <p>Change your password, or delete your account</p>
          </div>
          {
            this.state.message.map((msg) => {
              return (
                <div
                  key={ msg.length }
                  className="submit-messages">
                  <p>{ msg }</p>
                </div>
              );
            })
          }
          {
            this.state.isEditing
              ? (
                <div className="settings-fields">
                  <label htmlFor="password">Confirm your password</label>
                  <input
                    type="password"
                    id="password"
                    name="currentPassword"
                    onChange={ this.onChangeHandler }
                    disabled={ this.state.isLoading }
                  />
                  <p>Please enter your current password to proceed.</p>
                  <PlainBtn
                    onClick={ this.verifyCurrentPassword }>
                    {
                      this.state.isLoading
                        ? <LoadingCircle color="#fff" />
                        : 'Submit'
                    }
                  </PlainBtn>
                </div>
              )
              : null
          }
          {
            this.state.hasBeenVerified
              ? (
                <div className="settings-fields">
                  <label htmlFor="newPassword">Enter new password</label>
                  <input
                    type="password"
                    id="newPassword"
                    name="newPassword"
                    onChange={ this.onChangeHandler }
                    disabled={ this.state.isLoading }
                  />
                  <p>Please enter new password.</p>
                  <input
                    type="password"
                    id="confirmPassword"
                    name="confirmPassword"
                    onChange={ this.onChangeHandler }
                    disabled={ this.state.isLoading }
                  />
                  <p>Please confirm the new password.</p>
                  <PlainBtn
                    disabled={ this.state.isLoading }
                    onClick={ this.submitNewPassword }>
                    {
                      this.state.isLoading
                        ? <LoadingCircle color="#fff" />
                        : 'Submit'
                    }
                  </PlainBtn>
                </div>
              )
              : null
          }
          <div className="button-group">
            <button type="submit">
              Change Password
            </button>
            <button type="submit">
              Delete Account
            </button>
          </div>
        </form>
      </div>
    );
  }
}

export default SecuritySettings;
