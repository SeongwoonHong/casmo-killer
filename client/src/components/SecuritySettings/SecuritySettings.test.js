/* eslint react/jsx-boolean-value: 0 */
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';

import axios from 'axios';

import { validatePassword } from '@sharedUtils/inputValidators';

import SecuritySettings from './SecuritySettings';

describe('<SecuritySettings />', () => {

  let wrapper;
  let component;

  let mockPost;
  let mockPut;

  let setStateSpy;

  beforeEach(() => {

    wrapper = mount(
      <MemoryRouter>
        <SecuritySettings
          strategy="local" />
      </MemoryRouter>
    );

    component = wrapper.find(SecuritySettings).instance();

    mockPost = sinon.stub(axios, 'post');
    mockPut = sinon.stub(axios, 'put');

    setStateSpy = sinon.spy(SecuritySettings.prototype, 'setState');

  });

  afterEach(() => {
    mockPost.restore();
    mockPut.restore();
    setStateSpy.restore();
  });

  it('should render the component consistently', () => {
    expect(shallow(<SecuritySettings />)).toMatchSnapshot();
  });

  it('should set the editing state to true when the button is clicked', () => {

    const initialBtn = wrapper.find('.user-form-button').first();

    initialBtn.simulate('click');

    expect(initialBtn.text()).toEqual('Cancel');
    expect(setStateSpy.calledWith({ isEditing: true })).toEqual(true);
    expect(component.state.isEditing).toEqual(true);

  });

  it('should reset the component to its initial state when the cancel button is clicked', () => {

    const { initialState } = component;

    wrapper
      .find(SecuritySettings)
      .instance()
      .setState({
        isEditing: true,
        isVerified: true
      });

    wrapper.find('.user-form-button').first().simulate('click');

    expect(setStateSpy.calledWith(initialState)).toEqual(true);

  });

  it('should verify the current password first before submitting', () => {

    // TODO: not sure why it has to be bound to the instance, not the prototype, need to investigate more
    // const spy = sinon.spy(SecuritySettings.prototype, 'verifyCurrentPassword');
    const spy = sinon.spy(component, 'verifyCurrentPassword');

    component.onSubmitHandler();

    expect(setStateSpy.calledWith({
      isLoading: true,
      message: ''
    })).toEqual(true);

    expect(spy.calledOnce).toEqual(true);

  });

  it('should not request the verification if the currentPassword field is empty', async () => {

    component.verifyCurrentPassword();
    expect(component.state.message).toEqual('Please enter your password.');

  });

  it('should send the entered password to the server and resolve the response from the server', async () => {

    const responseData = { status: 204 };
    const promise = Promise.resolve(responseData);

    mockPost.callsFake(() => promise);

    component.setState({ currentPassword: 'testpassword' });
    component.verifyCurrentPassword();

    await promise;

    expect(component.state.isVerified).toEqual(true);
    expect(component.state.isLoading).toEqual(false);

  });

  it('should display correct message when the request to the server fails', async () => {

    const responseData = { status: 205 };
    const promise = Promise.resolve(responseData);

    mockPost.callsFake(() => promise);

    component.setState({ currentPassword: 'testpassword' });
    component.verifyCurrentPassword();

    await promise;

    expect(component.state.message).toEqual('Failed to communicate with the server.');

  });

  it('should not submit the new password if it is different from the confirmPassword', async () => {

    const testData = {
      newPassword: 'testpassword',
      confirmPassword: 'testPassword'
    };

    component.setState(testData);

    component.submitNewPassword();

    const message = await validatePassword(testData.newPassword, testData.confirmPassword);

    expect(component.state.isSuccess).toEqual(false);
    expect(component.state.message).toEqual(message);

  });

  it('should submit the new password and resolve the response from the server', async (done) => {

    const testData = {
      newPassword: 'testpassword',
      confirmPassword: 'testpassword'
    };

    const responseData = { status: 204 };
    const responseDataTwo = { status: 12 };

    const promise = Promise.resolve(responseData);
    const promiseTwo = Promise.resolve(responseDataTwo);

    mockPost.callsFake(() => promiseTwo);

    await promise;

    mockPut.callsFake(() => promise);

    component.setState(testData);

    component.submitNewPassword();

    await promiseTwo;

    setImmediate(() => {
      expect(setStateSpy.calledWith({
        isSuccess: true,
        currentPassword: '',
        newPassword: '',
        confirmPassword: '',
        message: 'Your password has been successfully updated.'
      })).toEqual(true);
      done();
    });

  });

});
