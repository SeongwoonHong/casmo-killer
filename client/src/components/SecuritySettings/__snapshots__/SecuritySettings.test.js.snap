// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SecuritySettings /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SecuritySettings />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "button": <div
        className="Security-settings__button-group"
>
        <button
                className="User__form__btn Security-settings__form__btn"
                disabled={true}
                onClick={[Function]}
                type="button"
        >
                Change Password
        </button>
        <Link
                className="User__form__btn Security-settings__form__btn"
                replace={false}
                to="/user/deactivate"
        >
                Deactivate Account
        </Link>
</div>,
      "children": Array [
        <FormMessage
          className=""
          message=""
          type="error"
/>,
        <UserInputField
          disabled={false}
          isVisible={false}
          message="Please enter your current password to proceed."
          name="currentPassword"
          onChange={[Function]}
          placeholder=""
          title="Current Password"
          type="password"
          value=""
>
          <button
                    className="User__form__btn Security-settings__field__btn"
                    type="submit"
          >
                    Submit
          </button>
</UserInputField>,
        <UserInputField
          disabled={false}
          isVisible={false}
          message="Enter new password"
          name="newPassword"
          onChange={[Function]}
          placeholder=""
          title="New Password"
          type="password"
          value=""
/>,
        <UserInputField
          disabled={false}
          isVisible={false}
          message="Confirm your password"
          name="confirmPassword"
          onChange={[Function]}
          placeholder=""
          title="Confirm Password"
          type="password"
          value=""
>
          <button
                    className="User__form__btn Security-settings__field__btn"
                    type="submit"
          >
                    Submit
          </button>
</UserInputField>,
      ],
      "className": "Security-settings",
      "disabled": false,
      "formMsg": "Change your password, or deactivate your account",
      "formTitle": "Security Settings",
      "icon": "",
      "isLoading": false,
      "onSubmit": [Function],
      "title": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "",
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <button
            className="User__form__btn Security-settings__field__btn"
            type="submit"
>
            Submit
</button>,
          "disabled": false,
          "isVisible": false,
          "message": "Please enter your current password to proceed.",
          "name": "currentPassword",
          "onChange": [Function],
          "placeholder": "",
          "title": "Current Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "className": "User__form__btn Security-settings__field__btn",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": false,
          "isVisible": false,
          "message": "Enter new password",
          "name": "newPassword",
          "onChange": [Function],
          "placeholder": "",
          "title": "New Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <button
            className="User__form__btn Security-settings__field__btn"
            type="submit"
>
            Submit
</button>,
          "disabled": false,
          "isVisible": false,
          "message": "Confirm your password",
          "name": "confirmPassword",
          "onChange": [Function],
          "placeholder": "",
          "title": "Confirm Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "className": "User__form__btn Security-settings__field__btn",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "button": <div
          className="Security-settings__button-group"
>
          <button
                    className="User__form__btn Security-settings__form__btn"
                    disabled={true}
                    onClick={[Function]}
                    type="button"
          >
                    Change Password
          </button>
          <Link
                    className="User__form__btn Security-settings__form__btn"
                    replace={false}
                    to="/user/deactivate"
          >
                    Deactivate Account
          </Link>
</div>,
        "children": Array [
          <FormMessage
            className=""
            message=""
            type="error"
/>,
          <UserInputField
            disabled={false}
            isVisible={false}
            message="Please enter your current password to proceed."
            name="currentPassword"
            onChange={[Function]}
            placeholder=""
            title="Current Password"
            type="password"
            value=""
>
            <button
                        className="User__form__btn Security-settings__field__btn"
                        type="submit"
            >
                        Submit
            </button>
</UserInputField>,
          <UserInputField
            disabled={false}
            isVisible={false}
            message="Enter new password"
            name="newPassword"
            onChange={[Function]}
            placeholder=""
            title="New Password"
            type="password"
            value=""
/>,
          <UserInputField
            disabled={false}
            isVisible={false}
            message="Confirm your password"
            name="confirmPassword"
            onChange={[Function]}
            placeholder=""
            title="Confirm Password"
            type="password"
            value=""
>
            <button
                        className="User__form__btn Security-settings__field__btn"
                        type="submit"
            >
                        Submit
            </button>
</UserInputField>,
        ],
        "className": "Security-settings",
        "disabled": false,
        "formMsg": "Change your password, or deactivate your account",
        "formTitle": "Security Settings",
        "icon": "",
        "isLoading": false,
        "onSubmit": [Function],
        "title": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <button
              className="User__form__btn Security-settings__field__btn"
              type="submit"
>
              Submit
</button>,
            "disabled": false,
            "isVisible": false,
            "message": "Please enter your current password to proceed.",
            "name": "currentPassword",
            "onChange": [Function],
            "placeholder": "",
            "title": "Current Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "User__form__btn Security-settings__field__btn",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": false,
            "isVisible": false,
            "message": "Enter new password",
            "name": "newPassword",
            "onChange": [Function],
            "placeholder": "",
            "title": "New Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <button
              className="User__form__btn Security-settings__field__btn"
              type="submit"
>
              Submit
</button>,
            "disabled": false,
            "isVisible": false,
            "message": "Confirm your password",
            "name": "confirmPassword",
            "onChange": [Function],
            "placeholder": "",
            "title": "Confirm Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "User__form__btn Security-settings__field__btn",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
